### DISCUSSION TITLE
Get version at runtime

### DISCUSSION THREAD

**Discussion by lonix1** [adamralph/minver#807] [Q&A] [answered]
I've been using MinVer in my cake scripts to generate versioned nuget packages. It works successfully.

Now I need the version of my console programme at runtime. I tried:
````cs
Assembly.GetEntryAssembly().GetName().Version.ToString()                     // 0.0.0.0
Assembly.GetExecutingAssembly().GetName().Version.ToString()                 // 0.0.0.0
typeof(Program).GetTypeInfo().Assembly.GetName().Version.ToString()          // 0.0.0.0
````

In all cases I get `0.0.0.0`. I get that for debug, release and publish builds.

My config includes this:
````cs
<PackageReference Include="MinVer" Version="4.1.0">
  <PrivateAssets>all</PrivateAssets>
  <IncludeAssets>runtime;build;native;contentfiles;analyzers</IncludeAssets>
</PackageReference>
````

Can someone spot my mistake? Thank you.

(I did read the docs, but couldn't find something applicable to my use case.)

**Comment by adamralph** [comment, marked as answer]
The [README](https://github.com/adamralph/minver#version-numbers) states that `AssemblyVersion` is set to `{MinVerMajor}.0.0.0`, so if you are in the 0.x version range, `AssemblyVersion` will be `0.0.0.0`.

The `Version` .NET SDK property, however, is propagated to `AssemblyInformationalVersion`, so you can use that. [For example](https://github.com/adamralph/bullseye/blob/5c8fa753ccd6a4c821dcebcf987e588c4a5d9c4f/Bullseye/Internal/AssemblyExtensions.cs#L1-L11):

```c#
public static string GetVersion(this Assembly assembly) =>
    assembly.GetCustomAttributes(false).OfType<AssemblyInformationalVersionAttribute>().FirstOrDefault()
        ?.InformationalVersion ?? "Unknown";
```

**Reply by lonix1** [reply]
Thanks Ralph, that gives me `0.3.0-alpha.0.5` which is what I was expecting.

For anyone else who wants to know:
````cs
// assembly version
// = 0.0.0.0
Assembly.GetExecutingAssembly()
  .GetName().Version?.ToString() ?? "unknown"   

// assembly file version
// = 0.3.0.0
Assembly.GetExecutingAssembly()
  .GetCustomAttribute<AssemblyFileVersionAttribute>()?.Version ?? "unknown"

// assembly informational version
// = 0.3.0-alpha.0.5
Assembly.GetExecutingAssembly()
  .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion ?? "unknown"
````
I don't think the nullable items will ever be null, but better safe than sorry. However, they could be null when creating an `Assembly` [dynamically](https://github.com/dotnet/runtime/issues/75228#issuecomment-1240163422).

"Assembly version", "assembly file version" and "assembly informational version" are [defined here](https://docs.microsoft.com/en-ca/dotnet/standard/library-guidance/versioning#version-numbers).
